### PR TITLE
sd-journal: Add sd_journal_handle_sigbus()

### PR DESCRIPTION
**Pull request by daandemeyer**
Let's allow sd-journal users to hook into the sigbus queue as well
by exposing sigbus_push() via the new sd_journal_handle_sigbus()
function.

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
Will write docs and such once we agree on the look of the new function

File: src/basic/sigbus.c [hunk 1]
```diff
@@ -25,9 +25,17 @@ static unsigned n_installed = 0;
...
+void sigbus_push(void *addr) {
+        unsigned long ul;
+        void *aligned;
+
         assert(addr);
 
+        ul = (unsigned long) addr;
+        ul = ul / page_size();
+        ul = ul * page_size();
+        aligned = (void*) ul;
```
Comment (yuwata) [inline src/basic/sigbus.c:37]: When you touch the code, then please use `PAGE_ALIGN()` defined in memory-util.h.

**Comment by poettering** [comment]
i am definitely on board with the goal (there's even a TODO list item asking for this, since about forever).

but i am not sure the current implementation with the queue is really so nice, since the queue can overrun.

I think we can simplify this greatly, because there's something i didn't realize originally: we can simply maintain a table of all mmaps we care about in regular malloc() memory, and the sigbus handler can access that and mark mmaps as fucked. I previously assumed we had to synchronize that via mutexes or so, but that wouldnt work because if a sigbus handler interrupts our main code and that one has the mutex it would deadlock. however here's the thing: we can just maintain a single atomic boolean to lock accesses. because our code is single threaded we can set this boolean to indicate to the sigbus handler "hey, we are modifying the mmap table, don't bother with it", and then the sigbus handler can just avoid any handling of it. because while we manage the mmap table we can be sure that our code won't access the actually mmaped pages, hence any sigbus that is triggered at that time cannot be ours.

so, before we expose this users i think we should clean that part up

**Comment by DaanDeMeyer** [comment]
@poettering Any chance you could do a PR for that? I'm wildly out of touch when it comes to atomics

**Pull request closed without merging**